File: iosApp/iosApp/Modules/CurrencySend/CurrencySendPresenter.swift
// Created by web3d4v on 06/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

enum CurrencySendPresenterEvent {
    case dismiss
    case addressChanged(to: String)
    case pasteAddress
    case saveAddress
    case selectCurrency
    case amountChanged(to: BigInt)
    case feeChanged(to: NetworkFee)
    case qrCodeScan
    case feeTapped
    case review
}

protocol CurrencySendPresenter: AnyObject {
    func present()
    func handle(_ event: CurrencySendPresenterEvent)
}

final class DefaultCurrencySendPresenter {
    private weak var view: CurrencySendView?
    private let wireframe: CurrencySendWireframe
    private let interactor: CurrencySendInteractor
    private let context: CurrencySendWireframeContext
    
    private var sendTapped = false
    private var address: String?
    private var currency: Currency!
    private var amount: BigInt?
    private var fee: NetworkFee?
    
    private var items = [CurrencySendViewModel.Item]()
    private var fees = [NetworkFee]()

    init(
        view: CurrencySendView,
        wireframe: CurrencySendWireframe,
        interactor: CurrencySendInteractor,
        context: CurrencySendWireframeContext
    ) {
        self.view = view
        self.interactor = interactor
        self.wireframe = wireframe
        self.context = context
        loadContext()
    }
}

extension DefaultCurrencySendPresenter: CurrencySendPresenter {

    func present() {
        updateView(
            with: [
                .address(
                    .init(
                        placeholder: Localized("networkAddressPicker.to.address.placeholder", context.network.name),
                        value: formattedAddress ?? address,
                        isValid: false,
                        becomeFirstResponder: true
                    )
                ),
                .token(
                    .init(
                        amount: nil,
                        symbolIconName: currency.iconName,
                        symbol: currency.symbol.uppercased(),
                        maxAmount: currencyBalance,
                        maxDecimals: currency.decimalsUInt,
                        fiatPrice: currency.fiatPrice,
                        shouldUpdateTextFields: false,
                        shouldBecomeFirstResponder: false,
                        networkName: context.network.name
                    )
                ),
                .send(
                    .init(
                        networkFeeViewModel: .init(
                            estimatedFee: estimatedFee(),
                            feeName: feeName()
                        ),
                        buttonState: .ready
                    )
                )
            ]
        )
    }

    func handle(_ event: CurrencySendPresenterEvent) {
        switch event {
        case .dismiss:
            wireframe.dismiss()
        case .qrCodeScan:
            view?.dismissKeyboard()
            wireframe.navigate(to: .qrCodeScan(network: context.network, onCompletion: onQRCodeScanned()))
        case .selectCurrency:
            wireframe.navigate(to: .selectCurrency(onCompletion: onCurrencySelected()))
        case .saveAddress:
            wireframe.navigate(to: .underConstructionAlert)
        case let .addressChanged(address):
            guard !isAddress(address: address, equalTo: context.address) else { return }
            if
                let currentAddress = self.address,
                let formattedAddress = formattedAddress,
                formattedAddress.hasPrefix(address),
                formattedAddress.count == (address.count + 1)
            {
                updateAddress(with: String(currentAddress.prefix(currentAddress.count - 1)))
            } else {
                let isValid = context.network.isValidAddress(input: address)
                updateView(
                    address: address,
                    shouldTokenBecomeFirstResponder: isValid
                )
            }
        case .pasteAddress:
            let clipboard = UIPasteboard.general.string ?? ""
            let isValid = context.network.isValidAddress(input: clipboard)
            guard isValid else { return }
            updateView(address: clipboard, shouldTokenBecomeFirstResponder: isValid)
        case let .amountChanged(amount):
            updateView(amount: amount, shouldTokenUpdateTextFields: false)
        case let .feeChanged(fee):
            self.fee = fee
            updateCTA()
        case .feeTapped:
            view?.presentFeePicker(with: fees)
        case .review:
            sendTapped = true
            let isValidAddress = context.network.isValidAddress(input: address ?? "")
            guard let address = address, isValidAddress  else {
                updateView(shouldAddressBecomeFirstResponder: true)
                return
            }
            let balance = interactor.balance(currency: currency, network: context.network)
            guard let amount = amount, balance >= amount, amount > .zero, let fee = fee else {
                updateView(shouldTokenBecomeFirstResponder: true)
                return
            }
            guard let walletAddress = interactor.walletAddress else { return }
            view?.dismissKeyboard()
            let context = ConfirmationWireframeContext.Send(
                data: .init(
                    network: context.network,
                    currency: currency,
                    amount: amount,
                    addressFrom: walletAddress,
                    addressTo: address,
                    networkFee: fee
                )
            )
            wireframe.navigate(to: .confirmSend(context: context))
        }
    }
}

private extension DefaultCurrencySendPresenter {
    
    var currencyBalance: BigInt {
        interactor.balance(currency: currency, network: context.network)
    }
    
    func isAddress(address: String, equalTo addressTo: String?) -> Bool {
        guard let addressTo = addressTo else { return false }
        let addressToCompare = Formatters.Companion.shared.networkAddress.format(
            address: addressTo, digits: 8, network: context.network
        )
        return address == addressToCompare
    }
    
    func loadContext() {
        address = context.address
        currency = context.currency ?? interactor.defaultCurrency(network: context.network)
        refreshFees()
    }
    
    func refreshFees() {
        fees = interactor.networkFees(network: context.network)
        guard fee == nil, !fees.isEmpty else { return }
        fee = fees[0]
    }
    
    func onCurrencySelected() -> (Currency) -> Void {
        {
            [weak self] currency in
            guard let self = self else { return }
            self.currency = currency
            let newBalance = self.interactor.balance(currency: currency, network: self.context.network)
            let newAmount = BigInt.min(left: self.amount ?? .zero, right: newBalance)
            self.updateToken(with: newAmount, shouldUpdateTextFields: true)
        }
    }
    
    func updateView(with items: [CurrencySendViewModel.Item]) {
        view?.update(
            with: .init(
                title: Localized("currencySend.title", currency.symbol.uppercased()),
                items: items
            )
        )
    }
    
    func onQRCodeScanned() -> (String) -> Void {
        { [weak self] address in self?.updateAddress(with: address) }
    }
    
    func updateView(
        address: String? = nil,
        shouldAddressBecomeFirstResponder: Bool = false,
        amount: BigInt? = nil,
        shouldTokenUpdateTextFields: Bool = false,
        shouldTokenBecomeFirstResponder: Bool = false
    ) {
        updateAddress(
            with: address ?? self.address ?? "",
            becomeFirstResponder: shouldAddressBecomeFirstResponder
        )
        updateToken(
            with: amount ?? self.amount ?? .zero,
            shouldUpdateTextFields: shouldTokenUpdateTextFields,
            shouldBecomeFirstResponder: shouldTokenBecomeFirstResponder
        )
        updateCTA()
    }
    
    func updateAddress(
        with address: String,
        becomeFirstResponder: Bool = false
    ) {
        if !address.contains("...") { self.address = address }
        let isValid = context.network.isValidAddress(input: self.address ?? "")
        updateView(
            with: [
                .address(
                    .init(
                        placeholder: Localized("networkAddressPicker.to.address.placeholder", context.network.name),
                        value: formattedAddress ?? address,
                        isValid: isValid,
                        becomeFirstResponder: becomeFirstResponder
                    )
                )
            ]
        )
    }
    
    var formattedAddress: String? {
        guard let address = address else { return nil }
        guard context.network.isValidAddress(input: address) else { return nil }
        return Formatters.Companion.shared.networkAddress.format(
            address: address, digits: 8, network: context.network
        )
    }

    func updateToken(
        with amount: BigInt,
        shouldUpdateTextFields: Bool,
        shouldBecomeFirstResponder: Bool = false
    ) {
        self.amount = amount
        updateView(
            with: [
                .token(
                    .init(
                        amount: amount,
                        symbolIconName: currency.iconName,
                        symbol: currency.symbol.uppercased(),
                        maxAmount: currencyBalance,
                        maxDecimals: currency.decimalsUInt,
                        fiatPrice: currency.fiatPrice,
                        shouldUpdateTextFields: shouldUpdateTextFields,
                        shouldBecomeFirstResponder: shouldBecomeFirstResponder,
                        networkName: context.network.name
                    )
                )
            ]
        )
    }
    
    func updateCTA() {
        let isValidAddress = context.network.isValidAddress(input: address ?? "")
        let buttonState: CurrencySendViewModel.Send.State
        if !sendTapped {
            buttonState = .ready
        } else if !isValidAddress {
            buttonState = .invalidDestination
        } else if (amount ?? .zero) == .zero {
            buttonState = .insufficientFunds
        } else if (amount ?? .zero) > currencyBalance {
            buttonState = .insufficientFunds
        } else if (amount ?? .zero) == .zero {
            buttonState = .enterFunds
        } else {
            buttonState = .ready
        }
        updateView(
            with: [
                .send(
                    .init(
                        networkFeeViewModel: .init(
                            estimatedFee: estimatedFee(),
                            feeName: feeName()
                        ),
                        buttonState: buttonState
                    )
                )
            ]
        )
    }
    
    func estimatedFee() -> [Formatters.Output] {
        guard let fee = fee else { return [Formatters.OutputNormal(value: "-")] }
        var outputFormat = Formatters.Companion.shared.currency.format(
            amount: fee.amount, currency: fee.currency, style: Formatters.StyleCustom(maxLength: 10)
        )
        let min: Double = Double(fee.seconds) / Double(60)
        var value = ""
        if min > 1 {
            value = " ~ \(min.toString(decimals: 0)) \(Localized("min"))"
        } else {
            value = " ~ \(fee.seconds) \(Localized("sec"))"
        }
        outputFormat.append(
            Formatters.OutputNormal(value: value)
        )
        return outputFormat
    }
    
    func feeName() -> String {
        guard let fee = fee else { return "-" }
        return fee.name
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencySend/CurrencySendPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.currencySend

import com.sonsofcrypto.web3lib.formatters.Formatters
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.NetworkFee
import com.sonsofcrypto.web3lib.utils.BigInt
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.common.viewModels.CurrencyAmountPickerViewModel
import com.sonsofcrypto.web3walletcore.common.viewModels.NetworkAddressPickerViewModel
import com.sonsofcrypto.web3walletcore.common.viewModels.NetworkFeeViewModel
import com.sonsofcrypto.web3walletcore.extensions.*
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencySend.CurrencySendViewModel.ButtonState.*
import com.sonsofcrypto.web3walletcore.modules.currencySend.CurrencySendWireframeDestination.*

sealed class CurrencySendPresenterEvent {
    object QrCodeScan: CurrencySendPresenterEvent()
    data class AddressChanged(val value: String): CurrencySendPresenterEvent()
    data class PasteAddress(val value: String): CurrencySendPresenterEvent()
    object SaveAddress: CurrencySendPresenterEvent()
    object SelectCurrency: CurrencySendPresenterEvent()
    data class AmountChanged(val value: BigInt): CurrencySendPresenterEvent()
    object NetworkFeeTapped: CurrencySendPresenterEvent()
    data class NetworkFeeChanged(val value: NetworkFee): CurrencySendPresenterEvent()
    object Review: CurrencySendPresenterEvent()
    object Dismiss: CurrencySendPresenterEvent()
}

interface CurrencySendPresenter {
    fun present()
    fun handle(event: CurrencySendPresenterEvent)
}

class DefaultCurrencySendPresenter(
    private val view: WeakRef<CurrencySendView>,
    private val wireframe: CurrencySendWireframe,
    private val interactor: CurrencySendInteractor,
    private val context: CurrencySendWireframeContext,
): CurrencySendPresenter {
    private var address: String? = context.address
    private var currency: Currency = context.currency ?: interactor.defaultCurrency(context.network)
    private var amount: BigInt? = null
    private var networkFees: List<NetworkFee> = interactor.networkFees(context.network)
    private var networkFee: NetworkFee? = networkFees.firstOrNull()
    private var sendTapped = false

    override fun present() { updateView(addressBecomeFirstResponder = true) }

    override fun handle(event: CurrencySendPresenterEvent) {
        when (event) {
            is CurrencySendPresenterEvent.QrCodeScan -> {
                view.get()?.dismissKeyboard()
                wireframe.navigate(QrCodeScan { address = it; updateView() })
            }
            is CurrencySendPresenterEvent.AddressChanged -> {
                if (isAddress(event.value, context.address)) return
                updateAddressIfNeeded(event.value)
                val isValidAddress = context.network.isValidAddress(address ?: "")
                updateView(currencyBecomeFirstResponder = isValidAddress)
            }
            is CurrencySendPresenterEvent.PasteAddress -> {
                val isValidAddress = context.network.isValidAddress(event.value)
                if (!isValidAddress) return
                address = event.value
                updateView(currencyBecomeFirstResponder = isValidAddress)
            }
            is CurrencySendPresenterEvent.SaveAddress -> wireframe.navigate(UnderConstructionAlert)
            is CurrencySendPresenterEvent.SelectCurrency -> {
                wireframe.navigate(
                    SelectCurrency {
                        currency = it
                        amount = BigInt.min(amount ?: BigInt.zero(), currencyBalance)
                        updateView(currencyUpdateTextField = true)
                    }
                )
            }
            is CurrencySendPresenterEvent.AmountChanged -> {
                amount = event.value
                updateView()
            }
            is CurrencySendPresenterEvent.NetworkFeeTapped -> {
                view.get()?.presentNetworkFeePicker(networkFees)
            }
            is CurrencySendPresenterEvent.NetworkFeeChanged -> {
                networkFee = event.value
                updateView()
            }
            is CurrencySendPresenterEvent.Review -> {
                sendTapped = true
                val context = confirmationWireframeSendContext() ?: return
                wireframe.navigate(ConfirmSend(context))
            }
            is CurrencySendPresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
        }
    }

    private fun confirmationWireframeSendContext(): ConfirmationWireframeContext.Send? {
        if (!context.network.isValidAddress(address ?: "")) {
            updateView(addressBecomeFirstResponder = true)
            return null
        }
        val amount = amount ?: BigInt.zero()
        if (currencyBalance.isLessThan(amount) || (amount == BigInt.zero())) {
            updateView(currencyBecomeFirstResponder = true)
            return null
        }
        val walletAddress = interactor.walletAddress ?: return null
        view.get()?.dismissKeyboard()
        val networkFee = networkFee ?: return null
        return ConfirmationWireframeContext.Send(
            ConfirmationWireframeContext.SendContext(
                context.network,
                currency,
                amount,
                walletAddress,
                address!!,
                networkFee
            )
        )
    }

    private fun updateAddressIfNeeded(value: String) =
        if (value.isEmpty()) {
            address = value
        } else if (formattedAddress.startsWith(value) &&
            formattedAddress.count() == value.length + 1) {
            address = address?.dropLast(1)
        } else if (
            formattedAddress != formattedAddress(value) &&
            (address?.length ?: 0) < value.length
        ) {
            address = value
        } else {
            // nothing
        }

    fun isAddress(address: String, addressTo: String?): Boolean {
        val addressTo = addressTo ?: return false
        val addressToCompare = Formatters.networkAddress.format(addressTo, 8, context.network)
        return address == addressToCompare
    }

    private fun updateView(
        addressBecomeFirstResponder: Boolean = false,
        currencyUpdateTextField: Boolean = false,
        currencyBecomeFirstResponder: Boolean = false,
    ) {
        view.get()?.update(
            viewModel(
                addressBecomeFirstResponder, currencyUpdateTextField, currencyBecomeFirstResponder,
            )
        )
    }

    private fun viewModel(
        addressBecomeFirstResponder: Boolean = false,
        currencyUpdateTextField: Boolean = false,
        currencyBecomeFirstResponder: Boolean = false,
    ): CurrencySendViewModel =
        CurrencySendViewModel(
            Localized("currencySend.title", currency.symbol.uppercase()),
            listOf(
                networkAddressViewModel(addressBecomeFirstResponder),
                currencyViewModel(currencyUpdateTextField, currencyBecomeFirstResponder),
                sendViewModel()
            )
        )

    private fun networkAddressViewModel(
        becomeFirstResponder: Boolean = false,
    ): CurrencySendViewModel.Item.Address {
        val data = NetworkAddressPickerViewModel(
            Localized("networkAddressPicker.to.address.placeholder", context.network.name),
            formattedAddress,
            context.network.isValidAddress(address ?: ""),
            becomeFirstResponder
        )
        return CurrencySendViewModel.Item.Address(data)
    }

    private val formattedAddress: String get() = formattedAddress(address)

    private fun formattedAddress(address: String?): String {
        val address = address ?: return ""
        if (!context.network.isValidAddress(address)) { return address }
        return Formatters.networkAddress.format(address, 8, context.network)
    }

    private fun currencyViewModel(
        updateTextField: Boolean = false,
        becomeFirstResponder: Boolean = false
    ): CurrencySendViewModel.Item.Currency {
        val data = CurrencyAmountPickerViewModel(
            amount,
            currency.iconName,
            currency.symbol.uppercase(),
            currencyBalance,
            currency.decimals(),
            interactor.fiatPrice(currency),
            updateTextField,
            becomeFirstResponder,
            context.network.name
        )
        return CurrencySendViewModel.Item.Currency(data)
    }

    private val currencyBalance: BigInt get() = interactor.balance(currency, context.network)

    private fun sendViewModel(): CurrencySendViewModel.Item.Send {
        val data = CurrencySendViewModel.SendViewModel(
            networkFeeViewModel(),
            buttonStateViewModel()
        )
        return CurrencySendViewModel.Item.Send(data)
    }

    private fun networkFeeViewModel(): NetworkFeeViewModel {
        val networkFee = networkFee ?: return emptyNetworkFeeViewModel()
        val mul = interactor.fiatPrice(networkFee.currency)
        return networkFee.toNetworkFeeViewModel(mul)
    }

    private fun emptyNetworkFeeViewModel(): NetworkFeeViewModel =
        NetworkFeeViewModel("-", listOf(), listOf(), listOf())

    private fun buttonStateViewModel(): CurrencySendViewModel.ButtonState =
        if (!sendTapped) { READY }
        else if (!context.network.isValidAddress(address ?: "")) { INVALID_DESTINATION }
        else if (zeroBalance) { INSUFFICIENT_FUNDS }
        else if (zeroAmount && positiveBalance) { ENTER_FUNDS }
        else if (currencyBalance.isLessThan((amount ?: BigInt.zero()))) { INSUFFICIENT_FUNDS }
        else { READY }

    private val zeroAmount: Boolean get() {
        val amount = amount ?: return true
        return amount == BigInt.zero()
    }
    private val zeroBalance: Boolean = (currencyBalance == BigInt.zero())
    private val positiveBalance: Boolean = currencyBalance.isGreaterThan(BigInt.zero())
}
